tests/microbenchmark/test_rw_tensix.cpp

```diff
@@ -0,0 +1,128 @@
+        });
+        wname.clear();
+    }
+    bench.render(ankerl::nanobench::templates::csv(), results_csv);
+}
+
+void ReadAllCores32Bytes(std::shared_ptr<tt_SiliconDevice> device, std::ofstream& results_csv) {
+    std::vector<uint32_t> readback_vec = {};
+    std::uint32_t address = l1_mem::address_map::DATA_BUFFER_SPACE_BASE;
+
+    ankerl::nanobench::Bench bench;
+    for(auto& core :  device->get_virtual_soc_descriptors().at(0).workers) {
+        std::stringstream rname;
+        rname << "Read from device core (" << core.x << ", " << core.y << ")";
+        bench.title("Read 32 bytes").unit("reads").minEpochIterations(50).output(nullptr).run(rname.str(), [&] {
+            device->read_from_device(readback_vec, tt_cxy_pair(0, core), address, 0x20, "");
+        });
+        rname.clear();
+    }
+    bench.render(ankerl::nanobench::templates::csv(), results_csv);
+}
+
+void Write32BytesRandomAddr(std::shared_ptr<tt_SiliconDevice> device, std::ofstream& results_csv) {
+    std::vector<uint32_t> vector_to_write = {0, 1, 2, 3, 4, 5, 6, 7};
+    std::vector<uint32_t> readback_vec = {};
+    std::uint32_t address;
+
+    ankerl::nanobench::Bench bench;
+    for(auto& core :  device->get_virtual_soc_descriptors().at(0).workers) {
+        address = generate_random_address(1<<20); // between 0 and 1MB
+        std::stringstream wname;
+        wname << "Write to device core (" << core.x << ", " << core.y << ") @ address " << std::hex << address;
+        bench.title("Write 32 bytes random address").unit("writes").minEpochIterations(50).output(nullptr).run(wname.str(), [&] {
+            device->write_to_device(vector_to_write, tt_cxy_pair(0, core), address, "");
+        });
+        wname.clear();
+    }
+    bench.render(ankerl::nanobench::templates::csv(), results_csv);
+}
+
+void Read32BytesRandomAddr(std::shared_ptr<tt_SiliconDevice> device, std::ofstream& results_csv) {
+    std::vector<uint32_t> readback_vec = {};
+    std::uint32_t address;
+    auto &core = device->get_virtual_soc_descriptors().at(0).workers.at(0);
+
+    ankerl::nanobench::Bench bench;
+    for(int i = 0; i < 100; i++) {
+        address = generate_random_address(1<<20); // between 0 and 1MB
+        std::stringstream rname;
+        rname << "Read from device core (" << core.x << ", " << core.y << ") @ address " << std::hex << address;
+        bench.title("Read 32 bytes random address").unit("reads").minEpochIterations(50).output(nullptr).run(rname.str(), [&] {
+            device->read_from_device(readback_vec, tt_cxy_pair(0, core), address, 0x20, "");
+        });
+        rname.clear();
+    }
+    bench.render(ankerl::nanobench::templates::csv(), results_csv);
+}
+
+int main(){
+    std::ofstream results_csv("ubench_results.csv", std::ios_base::app);
+
+    auto get_static_tlb_index = [] (tt_xy_pair target) {
+        int flat_index = target.y * 10 + target.x;  // grid_size_x = 10 for GS/WH ????? something is wrong here
```

Comment:
It would be really great if software that used UMD never had to worry about TLB indices.